app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,184 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    @names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !@names.kind_of?(Array) || @names.size > 20
+
+    if params[:topic_id].present?
+      @topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(@topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    @allowed_names = if params[:allowed_names].present?
+      raise Discourse::InvalidParameters(:allowed_names) if !params[:allowed_names].is_a?(Array)
+      params[:allowed_names] << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    @names.each do |name|
+      if user = users[name]
+        reason = user_reason(user)
+        user_reasons[name] = reason if reason.present?
+      elsif group = groups[name]
+        reason = group_reason(group)
+        group_reasons[name] = reason if reason.present?
+      end
+    end
+
+    if @topic && @names.include?(SiteSetting.here_mention) && guardian.can_mention_here?
+      here_count = PostAlerter.new.expand_here_mention(@topic.first_post, exclude_ids: [current_user.id]).size
+    end
+
+    serialized_groups = groups.values.reduce({}) do |hash, group|
+      serialized_group = { user_count: group.user_count }
+
+      if group_reasons[group.name] == :not_allowed &&
+          members_visible_group_ids.include?(group.id) &&
+          (@topic&.private_message? || @allowed_names.present?)
+
+        # Find users that are notified already because they have been invited
+        # directly or via a group
+        notified_count = GroupUser
+          # invited directly
+          .where(user_id: topic_allowed_user_ids)
+          .or(
+            # invited via a group
+            GroupUser.where(
+              user_id: GroupUser.where(group_id: topic_allowed_group_ids).select(:user_id)
+            )
+          )
+          .where(group_id: group.id)
+          .select(:user_id).distinct.count
+
+        if notified_count > 0
+          group_reasons[group.name] = :some_not_allowed
+          serialized_group[:notified_count] = notified_count
+        end
+      end
+
+      hash[group.name] = serialized_group
+      hash
+    end
+
+    render json: {
+      users: users.keys,
+      user_reasons: user_reasons.compact,
```

Comment:
```suggestion
      user_reasons: user_reasons,
```

I believe `compact` is no longer necessary here since we only push to the array when a reason is present.